compilers/concrete-compiler/compiler/lib/Dialect/FHE/Analysis/MANP.cpp

```diff
@@ -915,6 +950,120 @@ class MANPAnalysis
+    auto stride = iterCount;
+    for (size_t i = 0; i < loopRange.size(); i++) {
+      stride /= loopRange[i];
+      strides.push_back(stride);
+    }
+
+    // init block arguments' MANP: map block arguments with op operands
+    for (auto arg : genericOp.getBlock()->getArguments()) {
+      auto argIndex = arg.getArgNumber();
+      auto operandRange = genericOp.getInputs();
+      if (argIndex >= operandRange.size()) {
+        argIndex -= operandRange.size();
+        operandRange = genericOp.getOutputs();
+      }
+      valueToManp[arg] = getLatticeElement(operandRange[argIndex])->getValue();
+    }
+
+    // indices at a specific iteration
+    llvm::SmallVector<int64_t> indices(loopRange.size(), 0);
+    std::optional<llvm::APInt> result = {};
+    for (auto i = 0; i < iterCount; i++) {
+      // TODO: can be used to have a more precise MANP value (when dealing with
+      // constants)
+      // - MANP values for each element of the tensor
+      // - use constant values to estimate MANP (instead of conservative)
+      // - take the max MANP in a tensor at the end
+      for (size_t iterPos = 0; iterPos < indices.size(); iterPos++) {
+        indices[iterPos] = (i / strides[iterPos]) % loopRange[iterPos];
+      }
+      genericOp.getBody()->walk([&](mlir::Operation *op) {
+        // we use the yield operation to set the final MANP value of the
+        // operation, as well as to update the MANP of the block argument
+        // representing the output of the operation (when needed)
+        if (auto yieldOp = mlir::dyn_cast<mlir::linalg::YieldOp>(op)) {
+          auto manp = fetchOrFallbackToAnalysis(yieldOp->getOperand(0));
+          result = manp->getMANP().value();
+          // if the output map is constant, then we need to accumulate MANP
+          // values in the output argument by updating it with the yielded value
+          // MANP
+          if (isOutputMapConst) {
+            auto outputBlockArg = genericOp.getBlock()->getArgument(
+                genericOp.getBlock()->getNumArguments() - 1);
+            valueToManp[outputBlockArg] = *manp;
+          }
+        }
+        // compute using the op and operand manp values
+        mlir::SmallVector<const MANPLattice *> latticeOperands;
+        for (auto operand : op->getOperands()) {
+          auto lattice = new MANPLattice(operand);
+          lattice->join(*fetchOrFallbackToAnalysis(operand));
+          latticeOperands.push_back(lattice);
+        }
+        std::optional<llvm::APInt> norm2SqEquiv =
+            norm2SqEquivFromOp(op, latticeOperands);
+        // free space of lattice elements
+        for (auto toFree : latticeOperands) {
+          delete toFree;
+        }
+        // update the MANP of the result value
+        if (op->getNumResults() > 0) {
+          valueToManp[op->getResult(0).cast<mlir::Value>()] =
+              MANPLatticeValue(norm2SqEquiv);
+        } else {
+          valueToManp[op->getResult(0).cast<mlir::Value>()] =
```

Comment:
Will it be always possible to `op->getResult(0)` if `op->getNumResults() == 0`? What's the behaviour in this case?

Reply:
Oops! I don't think there should be an `else` in here at all